docs/source/generics.rst

```diff
@@ -489,46 +697,46 @@ Type variables with value restriction
 By default, a type variable can be replaced with any type. However, sometimes
 it's useful to have a type variable that can only have some specific types
 as its value. A typical example is a type variable that can only have values
-``str`` and ``bytes``:
+``str`` and ``bytes``. This lets us define a function that can concatenate
+two strings or bytes objects, but it can't be called with other argument
+types (Python 3.12 syntax):
 
 .. code-block:: python
 
-   from typing import TypeVar
+   def concat[S: (str, bytes)](x: S, y: S) -> S:
+       return x + y
 
-   AnyStr = TypeVar('AnyStr', str, bytes)
+   concat('a', 'b')    # Okay
+   concat(b'a', b'b')  # Okay
+   concat(1, 2)        # Error!
 
-This is actually such a common type variable that :py:data:`~typing.AnyStr` is
-defined in :py:mod:`typing` and we don't need to define it ourselves.
 
-We can use :py:data:`~typing.AnyStr` to define a function that can concatenate
-two strings or bytes objects, but it can't be called with other
-argument types:
+The same thing is also possibly using the legacy syntax (Python 3.11 or earlier):
 
 .. code-block:: python
 
-   from typing import AnyStr
+   from typing import TypeVar
+
+   AnyStr = TypeVar('AnyStr', str, bytes)
 
    def concat(x: AnyStr, y: AnyStr) -> AnyStr:
        return x + y
 
-   concat('a', 'b')    # Okay
-   concat(b'a', b'b')  # Okay
-   concat(1, 2)        # Error!
-
-Importantly, this is different from a union type, since combinations
-of ``str`` and ``bytes`` are not accepted:
+No matter which syntax you use, such a type variable is called a type variable
+with a value restriction. Importantly, this is different from a union type,
```

Comment:
At runtime we call this a TypeVar with "constraints"; maybe we should also switch to this terminology in mypy?

Reply:
Created an issue: https://github.com/python/mypy/issues/17828